### PR TITLE
Add dydx-v4-proto c++

### PR DESCRIPTION
**Pull request by MementoRC**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by MementoRC** [comment]
@conda-forge/help-c-cpp, Hello, ready for review

**Comment by MementoRC** [comment]
@xhochy Is it acceptable to have both `proto` and `client` into one recipe as in: https://github.com/conda-forge/staged-recipes/pull/27325?

Upstream only provides one source for both, but I can continue working on separating them as in this PR

File: recipes/dydxprotocol-v4-proto-cpp/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  # Reuiqres GCC >= 7
+  skip: True  # [win]
+
+requirements:
+  build:
+    - {{ compiler('c') }}  # [not win]
+    - {{ compiler('cxx') }}  # [not win]
+    - {{ stdlib("c") }}
+    # Using experimental m2w64 toolchain for now
+    - mingw-w64-ucrt-x86_64-gcc  # [win and x86_64]
```
Comment (xhochy) [inline recipes/dydxprotocol-v4-proto-cpp/meta.yaml:63]: These should be available using normal terms now. At least there is a UCRT migration that uses these compiler but the standard `compiler()` functions

Comment (MementoRC) [reply recipes/dydxprotocol-v4-proto-cpp/meta.yaml:63]: You lost me. I need this for gcc>7, the `blst` recipe that I am working on still show `5.3.0` for` compiler('m2w64_c')`

Comment (xhochy) [reply recipes/dydxprotocol-v4-proto-cpp/meta.yaml:63]: This is the migration I'm referring to https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/migrations/r-base44_and_m2w64-ucrt.yaml

Comment (MementoRC) [reply recipes/dydxprotocol-v4-proto-cpp/meta.yaml]: `compiler('m2w64_c')` still installs gcc 5.3

Comment (MementoRC) [reply recipes/dydxprotocol-v4-proto-cpp/meta.yaml]: Ok, got it with the `conda_build_config.yaml`

**Comment by MementoRC** [comment]
@xhochy Oddly, I don't seem to know how to use the environment, CC and CXX do not seem to be set and by default cmake picks clang (which I don't install ... confused)

**Comment by xhochy** [comment]
It's green now. Do you still need help or should I review?

**Comment by MementoRC** [comment]
I am tempted to say yes and yes.

Also, just so you know ahead of time, I do modify the upstream to separate proto and client and provide dynamic libraries

There are no tests for this lib, so, I am still debugging whether any changes are needed by using it in the `client` PR, there is something odd with protobuf across platforms.

**Comment by MementoRC** [comment]
@xhochy The missing symbols are solved and tested against the client. I think it is ready for review, I remove the versioned .so but kept the `outputs` since I feel it helps me filter the content of the package. Thank you

File: recipes/dydxprotocol-v4-proto-cpp/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+build:
+  number: 0
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ stdlib("c") }}
+    - cmake >=3.25
+    - libprotobuf >=4.22  # [build_platform == target_platform]
```
Comment (xhochy) [inline recipes/dydxprotocol-v4-proto-cpp/meta.yaml:62]: I'm confused. I would have expected the reverse condition.

File: recipes/dydxprotocol-v4-proto-cpp/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    - m2w64-binutils  # [win]
+    - ninja
+    - pkg-config
+    - zlib
+  host:
+    - catch2
+    - libabseil >=20240116
+    - libprotobuf >=4.22
```
Comment (xhochy) [inline recipes/dydxprotocol-v4-proto-cpp/meta.yaml:70]: Omit the pins to get get global pin
```suggestion
    - libabseil
    - libprotobuf
```

File: recipes/dydxprotocol-v4-proto-cpp/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - {{ stdlib("c") }}
+      host:
+        - {{ pin_subpackage('libdydx_v4_proto-headers') }}
+        - libabseil >=20240116
+        - libprotobuf >=4.22
+      run:
+        - {{ pin_subpackage('libdydx_v4_proto-headers') }}
+        - libprotobuf >=4.22  # [win]
```
Comment (xhochy) [inline recipes/dydxprotocol-v4-proto-cpp/meta.yaml:100]: This should come in via run_exports
```suggestion
```

File: recipes/dydxprotocol-v4-proto-cpp/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    - zlib
+  host:
+    - catch2
+    - libabseil >=20240116
+    - libprotobuf >=4.22
+
+outputs:
+  - name: libdydx_v4_proto
+    build:
+      ignore_run_exports:
```
Comment (xhochy) [inline recipes/dydxprotocol-v4-proto-cpp/meta.yaml:75]: Please use `ignore_run_exports_from`

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/dydxprotocol-v4-proto-cpp/meta.yaml (hunk 2 above)
Comment (MementoRC) [reply recipes/dydxprotocol-v4-proto-cpp/meta.yaml]: Right, I forgot why (if any logical reason) I tried that

**Review by xhochy — APPROVED** [review]

**Pull request merged**